sdk/core/azure-core/inc/azure/core/internal/io/filehandleholder.hpp

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief An internal FILE* Holder.
+ *
+ */
+#pragma once
+
+#include <cstdio>
+
+namespace Azure { namespace IO { namespace Internal {
+
+  /**
+   * @brief FileHandleHolder ensures native handle resource is released.
+   */
+  class FileHandleHolder {
+  private:
+    FILE* m_value = nullptr;
+
+  public:
+    explicit FileHandleHolder() noexcept : m_value(nullptr){};
+
+    ~FileHandleHolder() noexcept
+    {
+      if (m_value != nullptr)
+      {
+        fclose(m_value);
+        m_value = nullptr;
+      }
+    }
+
+    void operator=(FILE* p) noexcept
+    {
+      if (p != nullptr)
+      {
+        m_value = p;
+      }
+    }
```

Comment:
why do we need this? What's the difference between this one and the one generated by compiler?

Reply:
I got it. You define this function so that `m_fileStreamHolder = fopen(filename.c_str(), "rb");` can work.

But I don't think this is the right way to do it. Instead, you should define a constructor `FileHandleHolder(FILE*)`, and initlaize `m_fileStreamHolder` in member initializer list of `FileBodyStream`, like `FileBodyStream() : m_fileStreamHolder (fopen(filename.c_str(), "rb")) {}`